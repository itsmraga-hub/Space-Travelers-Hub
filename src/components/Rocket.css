.rocket-card {
    display: flex;
    align-items: center;
    justify-content: center;
    width: 100%;
    height: 100%;
    padding: 1rem;
    border-radius: 0.5rem;
    background-color: #fff;
    box-shadow: 0 0 0.5rem rgba(0, 0, 0, 0.1);
    transition: all 0.2s ease-in-out;
}

.rocket-card img {
    width: 50%;
    height: 20rem;
    object-fit: cover;
    border-radius: 0.5rem;
}

.rocket-info {
    display: flex;
    flex-direction: column;
    justify-content: space-between;
    width: 100%;
    height: 100%;
    padding: 1rem;
}