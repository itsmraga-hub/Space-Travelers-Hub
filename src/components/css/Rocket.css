.rocket-card {
    display: flex;
    align-items: center;
    justify-content: center;
    width: 100%;
    height: 100%;
    padding: 1rem;
    border-radius: 0.5rem;
    background-color: #fff;
    box-shadow: 0 0 0.5rem rgba(0, 0, 0, 0.1);
    transition: all 0.2s ease-in-out;
}

.rocket-card img {
    width: 50%;
    height: 20rem;
    object-fit: cover;
    border-radius: 0.5rem;
}

.rocket-info {
    display: flex;
    flex-direction: column;
    justify-content: space-between;
    width: 100%;
    height: 100%;
    padding: 1rem;
}

.reserved {
    width: 12rem;
    height: 4rem;
    border: none;
    border-radius: 0.2rem;
    background-color: #017bfe;
    color: #fff;
    font-size: 1.2rem;
}

.reserve {
    background-color: #1aa1b9;
    color: #fff;
    padding: 0.5rem 1rem;
    width: fit-content;
}

.cancel {
    width: 12rem;
    height: 4rem;
    border: none;
    border-radius: 0.2rem;
    background-color: gray;
    color: #fff;
    font-size: 1.2rem;
}