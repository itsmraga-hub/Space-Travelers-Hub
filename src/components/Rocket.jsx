import { useState } from 'react';
import PropTypes from 'prop-types';
import { useSelector, useDispatch } from 'react-redux';
import { reserveRocket, cancelRocket } from '../redux/rocketsReducer';
import './css/Rocket.css';

function Rocket({ rocket }) {
  const rockets = useSelector((state) => state.rockets.rockets);
  const {
    // eslint-disable-next-line camelcase
    id, rocket_name, description, flickr_images,
  } = rocket;
  // eslint-disable-next-line camelcase
  const [image] = useState(flickr_images[0]);
  const dispatch = useDispatch();

  const reserve = () => {
    dispatch(reserveRocket(id));
  };

  const cancel = () => {
    dispatch(cancelRocket(id));
  };

  // check if rocket is reserved
  const { reserved } = rockets.find((rocket) => rocket.id === id);

  return (
    <div className="rocket-card">
      {/* eslint-disable-next-line camelcase */}
      <img src={image} alt={rocket_name} />
      <div className="rocket-info">
        {/* eslint-disable-next-line camelcase */}
        <h2>{rocket_name}</h2>
        <>
          {reserved ? (<p className="reserve">Reserved</p>) : null}
          {description}
        </>
        {/* eslint-disable-next-line react/prop-types */}
        {!rocket.reserved && (
        <button type="button" className="reserved" onClick={reserve}>Reserve Rocket</button>
        )}
        {/* eslint-disable-next-line react/prop-types */}
        {rocket.reserved && (
        <button type="button" className="cancel" onClick={cancel}>Cancel Reservation</button>
        )}
      </div>
    </div>
  );
}

Rocket.propTypes = {
  rocket: PropTypes.shape({
    id: PropTypes.number.isRequired,
    rocket_name: PropTypes.string.isRequired,
    description: PropTypes.string.isRequired,
    flickr_images: PropTypes.arrayOf(PropTypes.string).isRequired,
  }).isRequired,
};

export default Rocket;
