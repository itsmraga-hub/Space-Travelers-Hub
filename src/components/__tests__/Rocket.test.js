import renderer from 'react-test-renderer';

import Rocket from '../Rocket';

// props id, rocket_name, description, image, reserve, cancel, isReserved,

const rocket = {
  id: 1,
  rocket_name: 'SpaceX',
  description: 'Lorem ipsum text generate',
  image: 'https://images2.imgbox.com/3c/0e/T8iJcSN3_o.png',
  reserve: () => true,
  cancel: () => true,
  reserved: false,
};

const rocket2 = {
  id: 2,
  rocket_name: 'SpaceX',
  description: 'Lorem ipsum text generate',
  image: 'https://images2.imgbox.com/3c/0e/T8iJcSN3_o.png',
  reserve: () => true,
  cancel: () => true,
  reserved: true,
};

it('<Rocket /> renders properly as expected', () => {
  const {
    id, rocket_name, description, image, reserve, cancel, reserved,
  } = rocket;
  const tree = renderer.create(<Rocket
    id={id}
    rocket_name={rocket_name}
    description={description}
    image={image}
    reserve={reserve}
    cancel={cancel}
    reserved={reserved}
  />).toJSON();
  expect(tree).toMatchSnapshot();
});

it('<Rocket /> renders properly as expected', () => {
  const {
    id, rocket_name, description, image, reserve, cancel, reserved,
  } = rocket2;
  const tree = renderer.create(<Rocket
    id={id}
    rocket_name={rocket_name}
    description={description}
    image={image}
    reserve={reserve}
    cancel={cancel}
    reserved={reserved}
  />).toJSON();
  expect(tree).toMatchSnapshot();
});
