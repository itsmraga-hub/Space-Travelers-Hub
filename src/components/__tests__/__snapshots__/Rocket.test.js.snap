// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Rocket /> renders properly as expected 1`] = `
<div
  className="rocket-card"
>
  <img
    alt="SpaceX"
    src="https://images2.imgbox.com/3c/0e/T8iJcSN3_o.png"
  />
  <div
    className="rocket-info"
  >
    <h2>
      SpaceX
    </h2>
    <p
      className="reserve"
    >
      Not reserved
    </p>
    Lorem ipsum text generate
    <div
      className="rocket-buttons"
    >
      <button
        className="reserved"
        onClick={[Function]}
        type="button"
      >
        Reserve Rocket
      </button>
    </div>
  </div>
</div>
`;

exports[`<Rocket /> renders properly as expected 2`] = `
<div
  className="rocket-card"
>
  <img
    alt="SpaceX"
    src="https://images2.imgbox.com/3c/0e/T8iJcSN3_o.png"
  />
  <div
    className="rocket-info"
  >
    <h2>
      SpaceX
    </h2>
    <p
      className="reserve"
    >
      Not reserved
    </p>
    Lorem ipsum text generate
    <div
      className="rocket-buttons"
    >
      <button
        className="reserved"
        onClick={[Function]}
        type="button"
      >
        Reserve Rocket
      </button>
    </div>
  </div>
</div>
`;
